Use spdlog for logging

theartful opened the pull request:
spdlog is a logging library that uses "fmt" for formatting (which is another great library that might get added to the standard!).
TODO:
- [x] Add spdlog to CMake
- [ ] Discuss what to do next!

Ideas:
1. Allow users to change 'sinks' (using spdlog terminology), so that loggers can write to different targets without exposing spdlog to them (?).
2. Enable each algorithm to have its own spd logger (=> SGIO object). Might require us to do something like RandomMixin so that it's children would write to the same logger.
3. Get rid of macros (would probably mean getting rid of line number and file info).

theartful committed:
Add spdlog to CMake

vigsterkr commented on src/shogun/io/SGIO.cpp:
```diff
@@ -41,19 +42,12 @@ const char* SGIO::message_strings_highlighted[NUM_LOG_LEVELS]={"[GCDEBUG] \0", "
-	print_message = &print_default;
-	print_warning = &print_default;
-	print_error = &print_default;
+	logger = spdlog::stdout_color_mt("console");
```
mmm maybe i would opt for the async logger....?

vigsterkr reviewed (commented):
i guess you'll need to wrap around this: https://github.com/gabime/spdlog/wiki/4.-Sinks#adding-sinks-to-the-logger-after-creation

to be able to add user defined sinks runtime...?

theartful committed:
Use spdlog in SGIO

We have two streams (stdout, stderr) that can be redirected to
any number of different sinks

theartful committed:
Drop io from SGObject

theartful committed:
Use fmt::ptr for printing pointers

this casts a pointer to void* since printing any other pointer types
is UB

theartful committed:
Replace printf specifiers

theartful committed:
Update interfaces

theartful committed:
Do not include spdlog in the header files

using spdlog as a header only library increases compile time greatly

theartful commented on src/shogun/io/SGIO.h:
```diff
@@ -42,13 +51,10 @@ enum EMessageType
+	MSG_WARN=3,
+	MSG_ERROR=4,
+	MSG_CRITICAL=5,
+	MSG_MESSAGEONLY=6
```
only 7 levels instead of 10

karlnapf replied:
yeah I actually never saw most of those used :)

theartful commented on src/shogun/io/SGIO.h:
```diff
@@ -220,6 +171,19 @@ struct substring
+	const char* file;
+	int32_t line;
+	const char* function;
+};
```
could've used spdlog::source_loc, but I chose not to include spdlog

theartful commented on src/shogun/io/SGIO.h:
```diff
@@ -328,89 +300,45 @@ class SGIO
-		/** set target to stdout */
-		inline void set_target_to_stdout() { set_target(stdout); }
+		/** redirects stderr to another sink */
+		void redirect_stderr(std::shared_ptr<spdlog::sinks::sink> sink);
```
there are only two sinks: one for stdout and the other for stderr. if the user wants to redirect to multiple sinks, he could use "dist_sink"

theartful commented on src/shogun/io/SGIO.h:
```diff
@@ -493,40 +415,16 @@ class SGIO
+		void message_(EMessageType prio, const SourceLocation& loc, const fmt::string_view& msg) const;
+
+		/** Updates log pattern */
+		void update_pattern();
```
pattern is `[date time {source_location} loglevel] msg`
example:
`[07/27/19 23:39:33 debug] Comparing parameter DynamicObjectArray::dim1_size of type int.`

theartful commented on src/shogun/io/SGIO.h:
```diff
@@ -535,39 +433,29 @@ class SGIO
+		class RedirectSink;
+		std::shared_ptr<RedirectSink> io_sink;
+		std::shared_ptr<spdlog::logger> io_logger;
+		std::shared_ptr<spdlog::details::thread_pool> thread_pool;
```
loggers hold `weak_ptr` of the thread_pool, so have to hold it in the class

theartful commented on src/shogun/io/SGIO.h:
```diff
@@ -535,39 +433,29 @@ class SGIO
+	{
+		fmt::memory_buffer msg;
+		fmt::format_to(msg, fmt, args...);
+		message_(prio, loc, fmt::string_view(msg.data(), msg.size()));
```
do the formatting in the header, and the logging in the cpp file. slightly inefficient because `should_log` is called twice

theartful commented on src/shogun/io/SGIO.cpp:
```diff
@@ -29,228 +37,188 @@
-
-
-void print_default(FILE* target, const char* str)
+class Formatter : public spdlog::formatter
```
This formatter is there just to support the loglevel "MSG_MESSAGEONLY" to not have any patterns

theartful commented on src/shogun/io/SGIO.h:
```diff
@@ -84,124 +87,72 @@ enum EMessageLocation
-#define SG_DEBUG(...) {												\
-	if (SG_UNLIKELY(io->loglevel_above(MSG_DEBUG)))					\
-		io->message(MSG_DEBUG, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);	\
+#define SG_GCDEBUG(...) {								\
```
I have no strong feelings about macros, so I left them

karlnapf replied:
changing them would require some pro/con list I guess?

theartful commented on src/shogun/io/SGIO.h:
```diff
@@ -84,124 +87,72 @@ enum EMessageLocation
-		SG_IO->template error<Exception>(                                      \
-		    MSG_ERROR, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);  \
-	}
-#define SG_SERROR(...) SG_STHROW(ShogunException, __VA_ARGS__)
```
no need for the "SG_S{logleve}" class of macros

karlnapf replied:
Can you briefly point me to why this is not needed anymore?

theartful replied:
SGObject's member variable "io" is removed, and only the global object is used "env()->io().

theartful committed:
Remove sg_print_functions.cpp

theartful committed:
run clang-format

karlnapf commented on src/shogun/io/SGIO.h:
```diff
@@ -84,124 +87,72 @@ enum EMessageLocation
-	}
-#define SG_UNSTABLE(func, ...) { io->message(MSG_WARN, __PRETTY_FUNCTION__, __FILE__, __LINE__, \
+
+#define SG_UNSTABLE(func, ...) { SG_IO->message(MSG_WARN, _SRC_LOC, \
```
I never liked this message .... maybe we can remove it and only merge tested/stable code? :D

theartful replied:
they're mainly used with default constructors for classes that shouldn't support them

karlnapf replied:
maybe we can kill them?
@vigsterkr @iglesias @lisitsyn

iglesias replied:
> they're mainly used with default constructors for classes that shouldn't support them

Hmm. Why are they defined?

iglesias replied:
> maybe we can kill them?
> @vigsterkr @iglesias @lisitsyn

It sounds good to me. Just stating the obvious here; let's get rid of the macro uses first and then let's remove the macro.

karlnapf replied:
or both at once? :)

theartful replied:
@iglesias We support default constructors for serialization and deserialization. Also for testing cloning.

karlnapf commented on src/shogun/io/SGIO.h:
```diff
@@ -231,34 +195,38 @@ class SGIO
+		 * @param queue_size size of the message queue
+		 * @param n_threads number of logging threads
+		 */
+		void init_default_logger(uint64_t queue_size=128, uint64_t n_threads=1);
```
why the n threads parameter? wouldnt we use the shogun number of threads?

theartful replied:
to point out that these are different threads

karlnapf replied:
could you give me some use cases for those threads?

theartful replied:
In general, if the logging code is slow, the performance will take a great hit. And in multithreaded code, it will get even slower because of the mutex locks. Async loggers solve this by having multiple logging queues in different threads so that logging calls are nonblocking.

karlnapf commented on src/shogun/io/SGIO.h:
```diff
@@ -328,89 +300,45 @@ class SGIO
+			    MSG_ERROR, loc,
+			    "This feature is only "
+			    "available if Shogun is built "
+			    "with GPL codes.\n");
```
the cmake error on this was recently updated by @sbrice .. maybe it should be updated here as well?

karlnapf commented on src/shogun/io/SGIO.h:
```diff
@@ -493,40 +415,16 @@ class SGIO
-		 *
-		 * @return reference count
-		 */
-		int32_t unref();
```
thank god this is gone :)

karlnapf reviewed (commented):
made some comments.

You mentioned some high level questions / ideas you had in the recent emails. Shall we discuss those here?

theartful committed:
Get rid of macros

ghost commented:
I got rid of macros, but I'm not sure if that's a good thing, as macros are more efficient. Using functions means that all the arguments have to be evaluated, whether the argument is going to be used or not.
This is expensive if the arguments are function calls, and when the function is used a lot, like `require`.

karlnapf commented:
Is there any way to quantify how much we are loosing when getting rid of macros?
I think before doing this for good, it would be good to have a pro/con list for macros/functions.
Thoughts?

vigsterkr commented:
lets have the ci fixed, i.e. interfaces and go from there...

ghost commented:
@karlnapf imo the only difference that would have an impact is the lazy evaluation of macros. Other than that, functions are heavily inlined by compilers, so they behave more or less like them. And macros have the disadvantage of polluting the symbol space, and might have unpredictable behaviour if not done correctly (actually there is such a case with the current implementation where you can't have them as one-line in if-else statement).
The performance would depend on the arguments being called, but I think the overall performance of any algorithm wouldn't get affected that much. Should I benchmark some algorithms before and after?

theartful committed:
Remove SG_DEBUG and ASSERT when compiled in Debug mode

theartful committed:
Use require instead of assert in CMahalanobisDistance::init

theartful commented on src/shogun/distance/MahalanobisDistance.cpp:
```diff
@@ -37,7 +37,7 @@ bool CMahalanobisDistance::init(CFeatures* l, CFeatures* r)
 {
 	// FIXME: See comments in
 	// https://github.com/shogun-toolbox/shogun/pull/4085#discussion_r166254024
-	ASSERT(CRealDistance::init(l, r));
```
With ASSERT being removed on Release build, there should be a refactor to replace the user facing ones with require.

karlnapf replied:
That’s good!
So just to make sure o understand: assert then will become something like in Java: it is ignored unless explicitly asked not to (ie debug mode build)?
That’s good as we can be more liberal with expensive checks.

theartful replied:
yup

vigsterkr commented:
@theartful mmm i'm not so sure what's happening with the windows build: https://dev.azure.com/shogunml/shogun/_build/results?buildId=2152&view=logs&jobId=9ffa40d6-0008-5f71-b4ad-2956145303e6&taskId=4c6a8d5c-d9f7-5d77-51f8-d43faa094742&lineStart=40569&lineEnd=40570&colStart=1&colEnd=1

vigsterkr commented:
or maybe it really runs out of time... and this is just the message of the cancellation. could you maybe change the maximum minutes for the windows ci job to 180 minutes?

ghost commented:
Yes, it runs out of time. I'm writing unit-tests, once finished, I will push with the increased maximum time for windows.

theartful committed:
Add unit tests

vigsterkr commented:
@theartful this is good to be merged, right?

ghost commented:
@vigsterkr One final issue: I noticed that spdlog adds a new line at the end of any message by default, however, this can be prevented using
```
#define SPDLOG_EOL ""
```
which I added in `SGIO.cpp`. If you prefer the default behaviour, I will remove this line and the new line character at the end of the current logs.

vigsterkr commented:
since the formatting of the logs are prefixed with time and other information i'd say let's keep the default behaviour of spdlog (append '\n' to the end of the log). this would basically mean that in case of log commands we would need to remove '\n' in order to not to have a double new line, right? this could be handled by a regex, or?

ghost commented:
I agree, and yes regex would do the job. I will leave `io::print` without the newline at the end, since it has no formatting.

vigsterkr commented:
sounds good! once you've pushed and CI passed i'll merge this! thnx for this patch!

theartful committed:
Increate windows CI timeout

theartful committed:
Drop new line character at the end of logs

vinx13 commented:
hey @theartful nice work! can you update the doc on wiki https://github.com/shogun-toolbox/shogun/wiki/Assertions , and put a bit details on your final blog

ghost commented:
@vinx13 sure, will do that.

karlnapf commented:
> hey @theartful nice work! can you update the doc on wiki https://github.com/shogun-toolbox/shogun/wiki/Assertions , and put a bit details on your final blog

I second that actually, might be worth going through the docs a bit and see what needs update ... (or at least remove stale parts in wiki/docs)
